core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -0,0 +1,172 @@
+    final static TimeValue DEFAULT_INDEX_LAG_MAX_WAIT = TimeValue.timeValueSeconds(30);
+
+    /** protects changes to all internal state and signals changes in {@link #checkpoint} */
+    final Object mutex = new Object();
+
+    /** each bits maps to a seqNo in round robin fashion. a set bit means the seqNo has been processed */
+    final FixedBitSet processedSeqNo;
+
+    /** value of {@link #SETTINGS_INDEX_LAG_THRESHOLD } */
+    final int indexLagThreshold;
+    /** value of {#link #SETTINGS_INDEX_LAG_THRESHOLD } */
+    final TimeValue indexLagMaxWait;
+
+
+    /** the next available seqNo - used for seqNo generation */
+    volatile long nextSeqNo = 0;
+
+    /** the current local checkpoint, i.e., all seqNo lower<= this number have been completed */
+    volatile long checkpoint = -1;
+
+    public LocalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        indexLagThreshold = indexSettings.getSettings().getAsInt(SETTINGS_INDEX_LAG_THRESHOLD, DEFAULT_INDEX_LAG_THRESHOLD);
+        indexLagMaxWait = indexSettings.getSettings().getAsTime(SETTINGS_INDEX_LAG_MAX_WAIT, DEFAULT_INDEX_LAG_MAX_WAIT);
+        processedSeqNo = new FixedBitSet(indexLagThreshold);
+
+    }
+
+    /**
+     * issue the next sequence number
+     *
+     * Note that this method can block to honour maximum indexing lag . See {@link #SETTINGS_INDEX_LAG_THRESHOLD }
+     **/
+    public long generateSeqNo() {
+        synchronized (mutex) {
+            // we have to keep checking when ensure capacity returns because it release the lock and nextSeqNo may change
+            while (hasCapacity(nextSeqNo) == false) {
+                ensureCapacity(nextSeqNo);
+            }
+            return nextSeqNo++;
+        }
+    }
+
+    /**
+     * marks the processing of the given seqNo have been completed
+     * Note that this method can block to honour maximum indexing lag . See {@link #SETTINGS_INDEX_LAG_THRESHOLD }
+     **/
+    public long markSeqNoAsCompleted(long seqNo) {
+        synchronized (mutex) {
+            // make sure we track highest seen seqNo
+            if (seqNo >= nextSeqNo) {
+                nextSeqNo = seqNo + 1;
+            }
+            if (seqNo <= checkpoint) {
+                // this is possible during recover where we might replay an op that was also replicated
+                return checkpoint;
+            }
+            // just to be safe (previous calls to generateSeqNo/markSeqNoAsStarted should ensure this is OK)
+            ensureCapacity(seqNo);
+            int offset = seqNoToOffset(seqNo);
+            processedSeqNo.set(offset);
+            if (seqNo == checkpoint + 1) {
+                do {
+                    // clear the flag as we are making it free for future operations. do se before we expose it
+                    // by moving the checkpoint
+                    processedSeqNo.clear(offset);
```

Comment:
Maybe just clear a range of bits with `FixedBitSet#(int, int)` instead of clearing bit by bit? The implementation looks to be more efficient and would just require care around the offset wrapping.